pkg/ring/tokens.go

```diff
@@ -0,0 +1,102 @@
+package ring
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+const (
+	// TokensVersion1 is the version is a simple list of tokens.
+	TokensVersion1 = 1
+)
+
+// Tokens is a simple list of tokens.
+type Tokens []uint32
+
+// StoreToFile stores the tokens in the given directory.
+func (t Tokens) StoreToFile(tokenFilePath string) (e error) {
+	if tokenFilePath == "" {
+		return errors.New("path is empty")
+	}
+	tmp := tokenFilePath + ".tmp"
+	f, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		return err
+	}
+
+	fileClosed := false
+	defer func() {
+		if !fileClosed {
```

Comment:
This looks overcomplicated to me. The `f.Close()` executed in the defer will just cover the case an error occurred before the function will call `f.Close()` at line `55`. In any case, we don't want to return the `f.Close()` but always the original error (you have a check but looks superfluos to me, cause there's no cause we want to return the error from `f.Close()`).

From my perspective, this logic could be simplified doing:
- Remove `fileClosed`
- Always call `f.Close()` in the defer and ignore its error

Reply:
This is actually correct logic for closing files. Ignoring `f.Close()` errors is wrong. However, if fileClosed is false, and we're closing in defer, then there must be already other error.

Reply:
Agree with both the comments. There should already be an error if the file is not closed yet.